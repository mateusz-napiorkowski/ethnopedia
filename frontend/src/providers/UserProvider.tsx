import React, { createContext, useContext, useState, useEffect } from "react"
import { jwtDecode } from "jwt-decode"

export interface JWT {
    exp: number;
    username: string;
    firstName: string;
    userId: string;
}

interface UserContextProps {
    isUserLoggedIn: boolean;
    firstName: string;
    userId: string;
    jwtToken: string;
    setUserData: (isLoggedIn: boolean, name: string, jwtToken: string, userId: string) => void;
}

export const UserContext = createContext<UserContextProps>({
    isUserLoggedIn: false,
    firstName: "",
    userId: "",
    jwtToken: "",
    setUserData: () => {
    },
})

export const useUser = () => useContext(UserContext)

interface UserProviderProps {
    children: React.ReactNode;
}

export const UserProvider: React.FC<UserProviderProps> = ({ children }) => {
    const [isUserLoggedIn, setIsUserLoggedIn] = useState(false)
    const [firstName, setFirstName] = useState("")
    const [userId, setUserId] = useState("")
    const [jwtToken, setJwtToken] = useState("")

    const setUserData = (isLoggedIn: boolean, name: string, jwtToken: string, userId: string) => {
        setIsUserLoggedIn(isLoggedIn)
        setFirstName(name)
        setUserId(userId)
        setJwtToken(jwtToken)
    }

    useEffect(() => {
        const token = localStorage.getItem("token")

        if (token) {
            try {
                const decodedToken = jwtDecode<JWT>(token)
                const currentTime = Date.now() / 1000
                if (decodedToken.exp > currentTime) {
                    setUserData(true, decodedToken.firstName, token, decodedToken.userId)
                } else {
                    localStorage.removeItem("token")
                    setUserData(false, "", "", "")
                }
            } catch (error) {
                console.error("Invalid token:", error)
                localStorage.removeItem("token")
                setUserData(false, "", "", "")
            }
        }
    }, [])

    return (
        <UserContext.Provider value={{ isUserLoggedIn, firstName, userId, jwtToken, setUserData }}>
            {children}
        </UserContext.Provider>
    )
}
