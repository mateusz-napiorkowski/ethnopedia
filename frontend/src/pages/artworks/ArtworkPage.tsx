import LoadingPage from "../LoadingPage"
import Navbar from "../../components/navbar/Navbar"
import Navigation from "../../components/Navigation"
import { useState } from "react"
import WarningPopup from "../WarningPopup"
import { deleteArtworks, getArtwork } from "../../api/artworks"
import { useMutation, useQuery, useQueryClient } from "react-query"
import { useNavigate, useParams } from "react-router-dom"
import { useUser } from "../../providers/UserProvider"
import ArtworkDetails from "../../components/artwork/ArtworkDetails"

const ArtworkPage = () => {
    const { jwtToken } = useUser()
    const queryClient = useQueryClient()
    const { artworkId } = useParams<string>()
    const navigate = useNavigate()
    const [showMore, setShowMore] = useState(false)
    const [showDeleteArtworkWarning, setShowDeleteArtworkWarning] = useState(false)

    const { data: fetchedData } = useQuery({
        queryKey: ["artwork", artworkId],
        queryFn: () => getArtwork(artworkId as string),
        enabled: !!artworkId,
    })

    const findCategoryValue = (artwork: any, categoryName: string) => {
        const foundCategory = artwork.categories.find((category: any) => category.name === categoryName);
        return foundCategory ? foundCategory.value : "";
    }

    const deleteArtworksMutation = useMutation(() => deleteArtworks([artworkId as string], jwtToken as string), {
        onSuccess: () => {
            queryClient.invalidateQueries("artwork")
            navigate(`/collections/${artworkData.collectionName}/artworks/`)
        }
    })

    if (!fetchedData)
        return (
            <div data-testid="loading-page-container">
                <LoadingPage />
            </div>
        )

    const artworkData = fetchedData.artwork
    return (
        <>
            <div data-testid="loaded-artwork-page-container">
                <Navbar />
                {showDeleteArtworkWarning &&
                    <WarningPopup 
                        onClose={() => setShowDeleteArtworkWarning(!showDeleteArtworkWarning)}
                        deleteSelected={() => deleteArtworksMutation.mutate()}
                        warningMessage={"Czy na pewno chcesz usunąć rekord?"}
                    />
                }
                <section className="p-2 sm:p-4">
                    <div className="mx-auto max-w-screen-xl lg:px-6">
                        <Navigation artworkTitle={findCategoryValue(artworkData, "Tytuł")}/>
                        <ArtworkDetails
                            Tytuł={findCategoryValue(artworkData, "Tytuł")}
                            Artyści={findCategoryValue(artworkData, "Artyści")}
                            Rok={findCategoryValue(artworkData, "Rok")}
                            collectionName={artworkData.collectionName}
                            detailsToShow={showMore ? artworkData : {}}
                            handleEditClick={() => navigate(`edit-artwork`, {state:{categories: fetchedData.artwork.categories}})}
                            setShowDeleteArtworkWarning={setShowDeleteArtworkWarning}
                        />
                        <button type="button" onClick={() => setShowMore(!showMore)}
                                className="mt-4 px-4 py-2 bg-blue-500 text-white hover:bg-blue-400 font-semibold border-none">
                            {showMore ? "Pokaż mniej" : "Pokaż więcej"}
                        </button>
                    </div>
                </section>
            </div>
        </>
    )
}
export default ArtworkPage