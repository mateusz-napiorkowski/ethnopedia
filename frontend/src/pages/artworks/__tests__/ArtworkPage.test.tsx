import '@testing-library/jest-dom';
import { render, waitFor } from '@testing-library/react'
import ArtworkPage from "../ArtworkPage"
import { QueryClient, QueryClientProvider } from "react-query";
import { MemoryRouter, Routes, Route } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { UserContext } from '../../../providers/UserProvider';

const mockUseNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
   ...jest.requireActual('react-router-dom') as any,
  useNavigate: () => mockUseNavigate,
}));

const mockGetArtwork = jest.fn()
const mockDeleteArtwork = jest.fn()
jest.mock('../../../api/artworks', () => ({
    getArtwork: () => mockGetArtwork(),
    deleteArtwork: (artworkId: string, jwtToken: string) => mockDeleteArtwork(artworkId, jwtToken)
}))

const queryClient = new QueryClient();
const user = userEvent.setup()

const collection = "example collection"
const artworkId = "670c2aecc29b79e5aaef1b9b"
const renderPage = (
    queryClient: QueryClient, 
    userContextProps: any = {
        isUserLoggedIn: false,
        firstName: "",
        userId: "",
        jwtToken: undefined,
        setUserData: jest.fn()
    },
    collection: string = "example collection",
    artworkId: string = "670c2aecc29b79e5aaef1b9b"
    ) => {
        return render(
            <UserContext.Provider value={ userContextProps }>
                <QueryClientProvider client={queryClient}>
                    <MemoryRouter initialEntries={[`/collections/${collection}/artworks/${artworkId}`]}>
                        <Routes>
                            <Route path="/collections/:collection/artworks/:artworkId" element={<ArtworkPage />}/>
                        </Routes>  
                    </MemoryRouter>
                </QueryClientProvider>    
            </UserContext.Provider>
        );
};

const jwtToken = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6InRlc3Rvd3kiLCJmaXJzdE5hbWUiOiJ0ZXN0b3d5IiwidXN"
    + "lcklkIjoiNjZiNjUwNmZiYjY0ZGYxNjVlOGE5Y2U2IiwiaWF0IjoxNzI0MTg0MTE0LCJleHAiOjE3MjUxODQxMTR9.fzHPaXFMzQTVUf9IdZ0G6oeiaecc"
    + "N-rDSjRS3kApqlA"

const loggedInUserContextProps = {
    isUserLoggedIn: true,
    firstName: "123",
    userId: "66b6506fbb64df165e8a9ce6",
    jwtToken: jwtToken,
    setUserData: jest.fn()
};

const artworkData = {
    "artwork": {
        "_id": "670c2aecc29b79e5aaef1b9b",
        "createdAt": "2024-11-05T20:14:22.883Z",
        "updatedAt": "2024-11-05T20:14:22.883Z",
        "__v": 0,
        "categories": [
            {
                "name": "Tytuł",
                "values": [
                    "testowy"
                ],
                "subcategories": [
                    {
                        "name": "Podtytuł",
                        "values": [
                            "testowy podtytuł"
                        ],
                        "subcategories": []
                    }
                ]
            },
            {
                "name": "Artyści",
                "values": [
                    "testowi"
                ],
                "subcategories": []
            },
            {
                "name": "Rok",
                "values": [
                    "966"
                ],
                "subcategories": [
                    {
                        "name": "Miesiąc",
                        "values": [
                            "Wrzesień"
                        ],
                        "subcategories": [
                            {
                                "name": "Dzień",
                                "values": [
                                    "13"
                                ],
                                "subcategories": []
                            }
                        ]
                    },
                    {
                        "name": "Pora roku",
                        "values": [
                            "lato"
                        ],
                        "subcategories": []
                    }
                ]
            }
        ],
        "collectionName": "example collection"
    }
}

const artworkDataWithoutCategories = {
    "artwork": {
        "_id": "670c2aecc29b79e5aaef1b9b",
        "createdAt": "2024-11-05T20:14:22.883Z",
        "updatedAt": "2024-11-05T20:14:22.883Z",
        "__v": 0,
        "categories": [
        ],
        "collectionName": "example collection"
    }
}

describe("ArtworkPage tests", () => {
    beforeEach(() => {
        jest.clearAllMocks();
        queryClient.clear();
    });

    it("should render loading state", () => {           
        const {queryByTestId} = renderPage(queryClient)
        
        expect(queryByTestId("loading-page-container")).toBeInTheDocument()
        expect(queryByTestId("loaded-artwork-page-container")).not.toBeInTheDocument()
    })

    it("should render artwork page component correctly after data is fetched from API when main categories are specified", async () => {
        mockGetArtwork.mockReturnValue(artworkData)
        const {getByTestId, queryByTestId} = renderPage(queryClient)

        await waitFor(() => getByTestId('loaded-artwork-page-container'))

        expect(queryByTestId("loading-page-container")).not.toBeInTheDocument()
        expect(getByTestId("main-categories-container")).toMatchSnapshot()
    })

    it("should render artwork page component correctly after data is fetched from API when main categories are unspecified", async () => {
        mockGetArtwork.mockReturnValue(artworkDataWithoutCategories)
        const {getByTestId, queryByTestId} = renderPage(queryClient)

        await waitFor(() => getByTestId('loaded-artwork-page-container'))

        expect(queryByTestId("loading-page-container")).not.toBeInTheDocument()
        expect(getByTestId("main-categories-container")).toMatchSnapshot()
    })

    it("should render disabled edit button when user is not logged in", async () => {
        mockGetArtwork.mockReturnValue(artworkData)
        const {getByTestId, getByRole} = renderPage(queryClient)

        await waitFor(() => getByTestId('loaded-artwork-page-container'))

        expect(getByRole("button", {
            name: /edytuj/i
        })).toBeDisabled();
    })

    it("should navigate to edit artwork page after enabled edit button is clicked", async () => {
        mockGetArtwork.mockReturnValue(artworkData)
        const {getByTestId, getByRole } = renderPage(queryClient, loggedInUserContextProps)

        await waitFor(() => getByTestId('loaded-artwork-page-container'))
        await user.click(getByRole("button", {
            name: /edytuj/i
        }))

        expect(mockUseNavigate).toHaveBeenCalledWith(
            "edit-artwork", 
            {state: {categories: artworkData.artwork.categories}}
        )
    })

    it("should render disabled delete button when user is not logged in", async () => {
        mockGetArtwork.mockReturnValue(artworkData)
        const {getByTestId, getByRole} = renderPage(queryClient)

        await waitFor(() => getByTestId('loaded-artwork-page-container'))

        expect(getByRole("button", {
            name: /usuń/i
        })).toBeDisabled();
    })

    it("should render delete warning popup after enabled delete button is clicked", async () => {
        mockGetArtwork.mockReturnValue(artworkData)
        const {getByTestId, getByRole, getByText} = renderPage(queryClient, loggedInUserContextProps)

        await waitFor(() => getByTestId('loaded-artwork-page-container'))
        await user.click(getByRole("button", {
            name: /usuń/i
        }))

        await expect(getByText(/czy na pewno chcesz usunąć rekord?/i)).toBeInTheDocument()
    })

    it("should not render delete warning popup after user clicks cancel button", async () => {
        mockGetArtwork.mockReturnValue(artworkData)
        const {getByTestId, getByRole, queryByText} = renderPage(queryClient, loggedInUserContextProps)

        await waitFor(() => getByTestId('loaded-artwork-page-container'))
        await user.click(getByRole("button", {
            name: /usuń/i
        }))
        await user.click(getByRole("button", {
            name: /anuluj/i
        }))

        await expect(queryByText(/czy na pewno chcesz usunąć rekord?/i)).not.toBeInTheDocument()
    })

    it("should not render delete warning popup after user clicks exit button", async () => {
        mockGetArtwork.mockReturnValue(artworkData)
        const {getByTestId, getByLabelText, getByRole, queryByText} = renderPage(queryClient, loggedInUserContextProps)

        await waitFor(() => getByTestId('loaded-artwork-page-container'))
        await user.click(getByRole("button", {
            name: /usuń/i
        }))
        await user.click(getByLabelText("exit"))

        await expect(queryByText(/czy na pewno chcesz usunąć rekord?/i)).not.toBeInTheDocument()
    })

    it("should call deleteArtwork with correct arguments and navigate to collection page" +
        " after delete button in delete warning popup is clicked", async () => {
        mockGetArtwork.mockReturnValue(artworkData)
        mockDeleteArtwork.mockImplementation(() => ({acknowledged: true, deletedCount: 1}))
        const {getByTestId, getByRole, getByLabelText } = renderPage(queryClient, loggedInUserContextProps)

        await waitFor(() => getByTestId('loaded-artwork-page-container'))
        await user.click(getByRole("button", {
            name: /usuń/i
        }))
        user.click(getByLabelText("confirm"))

        await waitFor(() => expect(mockDeleteArtwork).toHaveBeenCalledWith(artworkId, jwtToken))
        expect(mockUseNavigate).toHaveBeenCalledWith(`/collections/${collection}/artworks/`)      
    })

    it("should render categories list after show more button is clicked", async () => {
        mockGetArtwork.mockReturnValue(artworkData)
        const {getByTestId, getByRole} = renderPage(queryClient)

        await waitFor(() => getByTestId('loaded-artwork-page-container'))
        await user.click(getByRole("button", {
            name: /pokaż więcej/i
        }))

        expect(getByRole("button", {
            name: /pokaż mniej/i
        })).toBeInTheDocument()
        expect(getByTestId("details-list")).toBeInTheDocument()
    })

    it("should not render categories list after show less button is clicked", async () => {
        mockGetArtwork.mockReturnValue(artworkData)
        const {getByTestId, getByRole, queryByTestId} = renderPage(queryClient)

        await waitFor(() => getByTestId('loaded-artwork-page-container'))
        await user.click(getByRole("button", {
            name: /pokaż więcej/i
        }))
        await user.click(getByRole("button", {
            name: /pokaż mniej/i
        }))
        
        await expect(getByRole("button", {
            name: /pokaż więcej/i
        })).toBeInTheDocument()
        expect(queryByTestId("details-list")).not.toBeInTheDocument()
    })

    it("should render nested categories list correctly", async () => {
        mockGetArtwork.mockReturnValue(artworkData)
        const {getByTestId, getByRole} = renderPage(queryClient)

        await waitFor(() => getByTestId('loaded-artwork-page-container'))
        await user.click(getByRole("button", {
            name: /pokaż więcej/i
        }))

        expect(getByTestId("details-list")).toMatchSnapshot()
    })

    it("should render categories list correctly when there aren't any categories", async () => {
        mockGetArtwork.mockReturnValue(artworkDataWithoutCategories)
        const {getByTestId, getByRole} = renderPage(queryClient)

        await waitFor(() => getByTestId('loaded-artwork-page-container'))
        await user.click(getByRole("button", {
            name: /pokaż więcej/i
        }))

        expect(getByTestId("details-list")).toMatchSnapshot()
    })
})