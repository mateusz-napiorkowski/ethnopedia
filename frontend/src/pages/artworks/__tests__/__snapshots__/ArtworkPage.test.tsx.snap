// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArtworkPage tests should render artwork page component correctly after data is fetched from API when main categories are specified 1`] = `
<div
  class="mt-2 flex-2 "
  data-testid="main-categories-container"
>
  <h1
    class="text-4xl font-bold text-gray-800 dark:text-white mb-2"
  >
    testowy
  </h1>
  <p
    class="text-xl text-gray-600 dark:text-gray-400 mt-1"
  >
    Artyści: 
    testowi
  </p>
  <p
    class="text-lg text-gray-500 dark:text-gray-300 mt-1"
  >
    Rok: 
    966
  </p>
  <p
    class="text-lg text-gray-500 dark:text-gray-300 mt-1"
  >
    Kolekcja: 
    example collection
  </p>
</div>
`;

exports[`ArtworkPage tests should render artwork page component correctly after data is fetched from API when main categories are unspecified 1`] = `
<div
  class="mt-2 flex-2 "
  data-testid="main-categories-container"
>
  <h1
    class="text-4xl font-bold text-gray-800 dark:text-white mb-2"
  />
  <p
    class="text-xl text-gray-600 dark:text-gray-400 mt-1"
  >
    Artyści: 
  </p>
  <p
    class="text-lg text-gray-500 dark:text-gray-300 mt-1"
  >
    Rok: 
  </p>
  <p
    class="text-lg text-gray-500 dark:text-gray-300 mt-1"
  >
    Kolekcja: 
    example collection
  </p>
</div>
`;

exports[`ArtworkPage tests should render categories list correctly when there aren't any categories 1`] = `
<ul
  class="px-4 list-[circle]"
  data-testid="details-list"
/>
`;

exports[`ArtworkPage tests should render nested categories list correctly 1`] = `
<ul
  class="px-4 list-[circle]"
  data-testid="details-list"
>
  <li>
    Tytuł
    : 
    testowy
    <ul
      class="px-4 list-[square]"
    >
      <li>
        Podtytuł
        : 
        testowy podtytuł
      </li>
    </ul>
  </li>
  <li>
    Artyści
    : 
    testowi
    <ul
      class="px-4 list-[square]"
    />
  </li>
  <li>
    Rok
    : 
    966
    <ul
      class="px-4 list-[square]"
    >
      <li>
        Miesiąc
        : 
        Wrzesień
        <ul
          class="px-4 list-[disc]"
        >
          <li>
            Dzień
            : 
            13
          </li>
        </ul>
      </li>
      <li>
        Pora roku
        : 
        lato
      </li>
    </ul>
  </li>
</ul>
`;
